flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/debezium/dispatcher/EventDispatcherImpl.java

```diff
@@ -206,11 +212,103 @@ private Struct schemaChangeRecordValue(SchemaChangeEvent event) throws IOExcepti
             value.put(HISTORY_RECORD_FIELD, historyStr);
             return value;
         }
 
+        /**
+         * Rewrites the table name in the Source if needed to handle schema changes properly.
+         *
+         * <p>This method addresses a specific issue when renaming multiple tables within a single
+         * statement, such as: {@code RENAME TABLE customers TO customers_old, customers_copy TO
+         * customers;}.
+         *
+         * <p>In such cases, Debezium's {@link io.debezium.connector.mysql.MySqlDatabaseSchema}
+         * emits two separate change events:
+         *
+         * <ul>
+         *   <li>{@code RENAME TABLE customers TO customers_old}
+         *   <li>{@code RENAME TABLE customers_copy TO customers}
+         * </ul>
+         *
+         * <p>Both events share a table name of {@code customers, customers_old} in their source
+         * info, which includes multiple table IDs in a single string.
+         *
+         * <p>On the other hand, the {@code TableChanges.TableChange#id} correctly identifies the
+         * schema change:
+         *
+         * <ul>
+         *   <li>The change for {@code RENAME TABLE customers_copy TO customers} has the {@code
+         *       customers} ID.
+         *   <li>The change for {@code RENAME TABLE customers TO customers_old} is empty.
+         * </ul>
+         *
+         * <p>The problem arises because {@link
+         * org.apache.flink.cdc.connectors.mysql.debezium.reader.BinlogSplitReader} does not expect
+         * multiple table IDs in the source info. As a result, changes for tables defined by the
+         * table filter configuration (e.g., {@code customers}) may be filtered out unintentionally.
+         * This can lead to schema changes not being saved in the state, which is crucial for
+         * recovering the job from a snapshot.
+         *
+         * <p>To resolve this issue, this method:
+         *
+         * <ol>
+         *   <li>Checks if the source info contains multiple table names.
+         *   <li>Verifies if the {@code TableChange#id} matches one of the table names.
+         *   <li>Updates the source info with the correct table name that conforms to Flink CDC
+         *       expectations, ensuring the schema change is saved correctly.
+         * </ol>
+         *
+         * @param event the schema change event emitted by Debezium.
+         * @return the updated source info with the corrected table name if necessary.
+         */
+        private Struct rewriteTableNameIfNeeded(SchemaChangeEvent event) {
+            Struct sourceInfo = event.getSource();
+            String tableName = sourceInfo.getString(TABLE_NAME_KEY);
+            if (tableName == null || tableName.isEmpty()) {
+                return sourceInfo;
+            }
+
+            List<String> tableNames = parseTableNames(tableName);
+            if (2 <= tableNames.size() && event.getDdl().toLowerCase().startsWith("rename")) {
```

Comment:
Seems some other DDL events like `DROP TABLE A, B` also carry multiple tableIds in their `TABLE_NAME_KEY` field. Could they be rewritten by this method?

Reply:
The rewrite logic explicitly checks that the DDL starts with `"RENAME"` using this condition:

```java
if (...event.getDdl().toLowerCase().startsWith("rename"))
```

So, other DDL events like `DROP TABLE` won’t be affected by this method.

Reply:
I was thinking there might be more DDLs that manipulates multiple tables besides `RENAME TABLE`, and this method could be extended to rewrite them, too. Maybe it should be done later in another ticket :)

Reply:
That’s a great idea! I agree that the method could be extended to handle other multi-table DDLs in the future. For now, I'll keep the focus on `RENAME TABLE`.